### ISSUE TITLE
Bug on Multi GPU : probably Invalid initialization of optimizer

### ISSUE THREAD

**Issue by unrealwill** [apache/incubator-mxnet#7847]
## Environment info
Operating System:
Ubuntu 16.04
MXNet version:
0.11.0

The first "update" when on multi-gpu does something strange. It then works fine but currently it breaks the save points.

I have manage to reproduce with simple code.

```
import mxnet as mx
from mxnet.io import DataBatch,DataDesc
import numpy as np


def buildModel():
    data = mx.symbol.Variable('data')
    text = mx.symbol.Variable('label')
    dense = mx.symbol.FullyConnected( data ,name="fc1", num_hidden=3 )
    out = mx.symbol.SoftmaxOutput( dense,text,name="softmaxoutput")


    #SINGLE GPU WORKS
    #mod = mx.mod.Module(out, context=[mx.gpu(0)], data_names=["data"], label_names=["label"])
    #Multi GPU DOESN'T WORK
    mod = mx.mod.Module(out, context=[mx.gpu(0),mx.gpu(1)], data_names=["data"], label_names=["label"])
    return mod

def initModel(m,bs):
    m.bind(data_shapes=[DataDesc("data", (bs, 100))],
           label_shapes=[DataDesc("label", (bs,))], force_rebind=True)
    m.init_params()
    m.init_optimizer(optimizer='adam', optimizer_params=[('learning_rate', 1e-3) ])

def demo():
    bs = 8
    m = buildModel()
    initModel(m, bs)

    target =  np.mod (np.arange(bs,dtype="int32"),3)
    x = np.random.randn(bs,100)

    for i in range( 10000 ):
        m.forward_backward(DataBatch( data= [mx.nd.array(x )] , label= [mx.nd.array(target)]) )
        res = m.get_outputs()
        m.update()
        if i == 9999:
            print(res[0].asnumpy())

    m.save_params("testmultigpu.params")

    print( " saving model ")

    print(" loading model ")
    mod = buildModel()
    initModel(mod, bs)
    mod.load_params("testmultigpu.params")
    for i in range( 2 ):
        mod.forward_backward(DataBatch( data= [mx.nd.array(x )] , label= [mx.nd.array(target)]) )
        res = mod.get_outputs()
        mod.update()
        print(res[0].asnumpy())


if __name__ == '__main__':
    demo()

```

On single GPU (on either GPU 0 or GPU 1 ) it work as expected :

> [[  9.99995232e-01   2.44983039e-06   2.24760834e-06]
>  [  3.24456596e-06   9.99994159e-01   2.64155938e-06]
>  [  2.56306998e-06   4.04469984e-06   9.99993324e-01]
>  [  9.99995232e-01   3.21645393e-06   1.59501099e-06]
>  [  3.40141946e-06   9.99994636e-01   2.01853436e-06]
>  [  2.76782998e-06   2.76159153e-06   9.99994516e-01]
>  [  9.99993443e-01   4.52491850e-06   1.98577322e-06]
>  [  2.69583552e-06   9.99994874e-01   2.39593828e-06]]
>  saving model 
>  loading model 
> [[  9.99995232e-01   2.44855755e-06   2.24643600e-06]
>  [  3.24292637e-06   9.99994159e-01   2.64023197e-06]
>  [  2.56174599e-06   4.04262528e-06   9.99993443e-01]
>  [  9.99995232e-01   3.21489892e-06   1.59423223e-06]
>  [  3.39969392e-06   9.99994636e-01   2.01750868e-06]
>  [  2.76647324e-06   2.76022502e-06   9.99994516e-01]
>  [  9.99993443e-01   4.52262339e-06   1.98476414e-06]
>  [  2.69446036e-06   9.99994874e-01   2.39472320e-06]]
> [[  9.99995470e-01   2.32505658e-06   2.12607233e-06]
>  [  3.09917482e-06   9.99994397e-01   2.51451502e-06]
>  [  2.44805597e-06   3.89877187e-06   9.99993563e-01]
>  [  9.99995351e-01   3.07445248e-06   1.50818892e-06]
>  [  3.24788380e-06   9.99994874e-01   1.90605579e-06]
>  [  2.64144001e-06   2.62431058e-06   9.99994755e-01]
>  [  9.99993682e-01   4.37004564e-06   1.89418438e-06]
>  [  2.57322654e-06   9.99995112e-01   2.27452938e-06]]

On multi GPU, it bugs at the first update (and lose all progress): 

> [[  9.99994874e-01   2.76984565e-06   2.39504266e-06]
>  [  3.15983220e-06   9.99994040e-01   2.84408338e-06]
>  [  2.57450711e-06   3.42540147e-06   9.99993920e-01]
>  [  9.99993920e-01   3.62951528e-06   2.51919687e-06]
>  [  3.75006653e-06   9.99994516e-01   1.67898190e-06]
>  [  3.70834505e-06   2.99744420e-06   9.99993324e-01]
>  [  9.99995351e-01   2.63477455e-06   1.99476813e-06]
>  [  2.33333003e-06   9.99994159e-01   3.47498735e-06]]
>  saving model 
>  loading model 
> [[  9.99994874e-01   2.76847504e-06   2.39385986e-06]
>  [  3.15823559e-06   9.99994040e-01   2.84264092e-06]
>  [  2.57316196e-06   3.42360840e-06   9.99993920e-01]
>  [  9.99993920e-01   3.62769856e-06   2.51791903e-06]
>  [  3.74813931e-06   9.99994516e-01   1.67812061e-06]
>  [  3.70651719e-06   2.99598082e-06   9.99993324e-01]
>  [  9.99995351e-01   2.63343804e-06   1.99374676e-06]
>  [  2.33214200e-06   9.99994159e-01   3.47323476e-06]]
> [[ 0.33770946  0.32504591  0.33724463]
>  [ 0.31802508  0.34699419  0.3349807 ]
>  [ 0.32322595  0.33247185  0.34430221]
>  [ 0.3667697   0.32238626  0.310844  ]
>  [ 0.34820592  0.3265925   0.3252016 ]
>  [ 0.32034403  0.33263206  0.34702393]
>  [ 0.36623162  0.33518958  0.2985788 ]
>  [ 0.34777871  0.32446057  0.32776073]]

**Comment by unrealwill** [comment]
I found a workaround to my problem : 
Adding after mod.load_params("testmultigpu.params")  : 
 // We force a proper initialization of the optimizer after the reloading of the parameter
mod.init_optimizer(optimizer='adam', optimizer_params=[('learning_rate', 1e-3)], force_init=True)

solves the issue (but the different behaviour between single and multi gpu is misleading).

**Comment by piiswrong** [comment]
@szha

**Comment by szha** [comment]
Looked into this for a while. The fundamental difference between these two modes are that single GPU one is not updated on kvstore while two-GPU version uses kvstore for weight update. The hypotheses on the problem:
1. optimizer state is not properly set (disproved by switching to SGD and still got similar effect)
2. the kvstore has some states on it that is not cleaned.

I will look into this more tomorrow with help from @piiswrong and @eric-haibin-lin.

**Comment by szha** [comment]
So the problem is that when kvstore is used, it needs to be bound to the parameters for updates. This means, load_params should be called before init_optimizer is called. The reason it works on one GPU is that kvstore isn't used, and parameter updates doesn't involve pushing and pulling.

So the fix is
```python
def initModel(m,bs, param_name=None):
    m.bind(data_shapes=[DataDesc("data", (bs, 100))],
           label_shapes=[DataDesc("label", (bs,))], force_rebind=True)
    m.init_params()
    if param_name:
        m.load_params(param_name)
    m.init_optimizer(optimizer='adam', optimizer_params=[('learning_rate', 1e-3) ])
...
# in demo
    print(" loading model ")
    mod = buildModel()
    initModel(mod, bs, "testmultigpu.params")
```

**Comment by unrealwill** [comment]
Ok thanks, I think I get it now.
I thought load_params was assigning the weight value to the already existing weight, but your remark seems to imply that it instead scratch_and_load_params to a fresh new set of params, which is why you would need to re_initialize the optimizer.

Coming from Keras, this feels unintuitive and error prone.

**Comment by szha** [comment]
@unrealwill yeah, the module is built for convenience, and sometimes the encapsulation can be leaky

**Comment by eric-haibin-lin** [comment]
I'll update module to throw proper warning/error messages for such case

**Comment by szha** [comment]
#8054 prevents the edge case from happening. However, I agree with @piiswrong that "load_param should work even after optimizer is initialized. This is a bug that needs to be fixed". Our forces are spreading thin right now and we really welcome contribution on patching this.

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.
